tools/fft.c

```diff
@@ -0,0 +1,381 @@
+	}
+
+
+	channel->sample_index++;
+	return 0;
+}
+
+int init_channel(struct channel_data *channel,
+		const bl_msg_channel_conf_t *msg, uint32_t window_count)
+{
+	memset(channel, 0, sizeof(*channel));
+	channel->channel = msg->channel;
+
+	/* I have decided to always use 50% overlap
+	 * If the window length in samples is even-numbered, there'd be 1
+	 * partly-full window when we have window_count full windows.
+	 * With odd-numbered window length, there's a second newly-started window
+	 */
+	channel->windows_capacity = window_count + 2;
+	channel->welch_window_count = window_count;
+	channel->windows = pfifo_create(channel->windows_capacity);
+	if (channel->windows == NULL) {
+		return -errno;
+	}
+	return 0;
+}
+
+int init_channel_samples(struct channel_data *channel,
+		uint32_t window_length_samples)
+{
+	struct sample_window *window;
+	channel->window_length = window_length_samples;
+	channel->new_window_interval = window_length_samples / 2;
+	if (!channel->windows) {
+		fprintf(stderr, "Failed to initialise channel. "
+				"Perhaps a channel config message is missing from the input "
+				"stream\n");
+		return -1;
+	}
+	window = create_sample_window(channel);
+	if (window == NULL) {
+		return -errno;
+	}
+	if (!pfifo_write(channel->windows, window)) {
+		return -1;
+	}
+	return 0;
+}
+
+void destroy_channel(struct channel_data *channel)
+{
+	struct sample_window *window;
+	if (channel->welch_output != NULL) {
+		free(channel->welch_output);
+	}
+	// Destroy all existing windows
+	while(pfifo_read(channel->windows, (void**) &window)) {
+		destroy_sample_window(window);
+	}
+	pfifo_destroy(channel->windows);
+}
+
+int read_stream(uint32_t window_length, uint16_t window_count)
+{
+	union bl_msg_data msg; // message for reading into
+	struct channel_data channels[BL_ACQ__SRC_COUNT]; //One per possible channel
+	struct channel_data *channel;
+
+	while (!bl_sig_killed && bl_msg_yaml_parse(stdin, &msg)) {
+		int ret;
+		uint32_t length_samples;
+		switch(msg.type) {
+		case BL_MSG_CHANNEL_CONF:
+			ret = init_channel(channels + msg.channel_conf.channel,
+					&msg.channel_conf, window_count);
+			if (ret < 0) {
+				return ret;
+			}
+			break;
+		case BL_MSG_START:
+			length_samples = window_length *
+				msg.start.frequency	/ 1000; // Magical 1000 from milliseconds.
+
+			// Create all the channels' buffers now we know the size
+			for (unsigned i = 0; i < BL_ACQ__SRC_COUNT; i++) {
+				ret = init_channel_samples(channels + i, length_samples);
+				if (ret < 0) {
+					return ret;
+				}
+			}
+			break;
+		case BL_MSG_SAMPLE_DATA16:
+		case BL_MSG_SAMPLE_DATA32:
+			channel = channels + msg.sample_data.channel;
+
+			for (unsigned i = 0; i < msg.sample_data.count; i++) {
+				if (msg.type == BL_MSG_SAMPLE_DATA16) {
+					add_sample_to_channel(channel,
+							(double) msg.sample_data.data16[i]);
+				} else {
+					add_sample_to_channel(channel,
+							(double) msg.sample_data.data32[i]);
+				}
+			}
+			break;
+		}
+	}
+	return 0;
```

Comment:
Channel cleanup, something like:

```suggestion
	ret = 0;
cleanup:
	for (unsigned i = 0; i < BL_ACQ_CHANNEL_COUNT; i++) {
		destroy_channel(channels + i);
	}
	return ret;
```